### PR TITLE
TCOMP-234: added SetupRuntime and SetupComponent

### PR DESCRIPTION
**Pull request by chmyga**
Extracted initialize and validate methods to Runtime interface.
Added SetupRuntime and SetupComponent interfaces.
Extended SourceOrSink interface from Runtime interface.
Added setup method to SetupRuntime without any arguments, because they will be passed to initialize method of parent Runtime interface.

### CODE REVIEW

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/275/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/275/)

File: components-api/src/main/java/org/talend/components/api/component/runtime/Runtime.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import org.talend.components.api.container.RuntimeContainer;
+import org.talend.components.api.properties.ComponentProperties;
+import org.talend.daikon.properties.ValidationResult;
+
+/**
+ * Base interface for defining properties and methods common to Runtime interfaces, like {@link SourceOrSink} or
+ * {@link SetupRuntime}.
+ */
+public interface Runtime {
+
```
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/runtime/Runtime.java]: I think the name "Runtime" it too easily confused with the java.lang.runtime. Can we pick a different name?

File: components-api/src/main/java/org/talend/components/api/component/runtime/SetupRuntime.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * Basic interface for defining properties and common methods required for Runtimes of components without connectors, or
+ * components which need some pre-processing before performing the actual job.
+ */
+public interface SetupRuntime extends Runtime {
+
+    /**
+     * Perform processes required before main Component's processes or in case a Component doesn't have any connections.
+     */
+    void setup();
+
```
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/runtime/SetupRuntime.java]: I don't understand why we need this interface. I think Runtime.validate() could be used for the same purpose, as that's what it's used for with the I/O components, so it would make it more uniform. It also gives you the ValidationResult which can nicely convey a failure.

File: components-api/src/main/java/org/talend/components/api/component/SetupComponent.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * Definition of a component which has Setup Runtime. It could be a component which doesn't have any input or output
+ * connectors, or a component which requires some pre-processing.
+ */
+public interface SetupComponent {
+
+    /**
+     * Return the {@link SetupRuntime}.
+     */
+    SetupRuntime getSetupRuntime();
+
```
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/SetupComponent.java]: Note sure about this interface. Whatever the current mechanism is to get the SourceOrSink should just be generalized to return the "Runtime" interface. No need for this extra method I think.

File: components-api/src/main/java/org/talend/components/api/component/runtime/Runtime.java (hunk 1 above)
Comment (chmyga) [inline components-api/src/main/java/org/talend/components/api/component/runtime/Runtime.java]: Sure, the name is not  that critical. I would be very thankful If you give any suggestions.

File: components-api/src/main/java/org/talend/components/api/component/runtime/SetupRuntime.java (hunk 2 above)
Comment (chmyga) [inline components-api/src/main/java/org/talend/components/api/component/runtime/SetupRuntime.java]: Doing some real work in validate() method is not the best idea, I think. validate should, actually, validate something, make sure that everything is correct. The name of the validate() method doesn't correspond to the method which does some real work.

File: components-api/src/main/java/org/talend/components/api/component/SetupComponent.java (hunk 3 above)
Comment (chmyga) [inline components-api/src/main/java/org/talend/components/api/component/SetupComponent.java]: This will be implemented by *ComponentDefinition classes, not SourceOrSink. Also, we assumed that components could be Input, Output, Endpoint and Setup at the same time. So getRuntime() method won't be the best way to call the methods returning runtimes. But it's all should be discussed for sure. I think that Sebastien could give some of his thoughts about this point also.

File: components-api/src/main/java/org/talend/components/api/component/runtime/Runtime.java (hunk 1 above)
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/runtime/Runtime.java]: How about ComponentRuntime?

File: components-api/src/main/java/org/talend/components/api/component/runtime/SetupRuntime.java (hunk 2 above)
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/runtime/SetupRuntime.java]: Sorry, I meant initialize() should handle it, not validate().

File: components-api/src/main/java/org/talend/components/api/component/SetupComponent.java (hunk 3 above)
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/SetupComponent.java]: Let's add getRuntime() in ComponentDefinition and have that return a ComponentRuntime object. I think that the InputComponentDefinition can still override that interface and not have to change. Seb is working on a related change that requires a modification to a top-level getRuntime() in ComponentDefinition (I think he needs to take the properties). So we should combine those ideas.

File: components-api/src/main/java/org/talend/components/api/component/runtime/SetupRuntime.java (hunk 2 above)
Comment (chmyga) [inline components-api/src/main/java/org/talend/components/api/component/runtime/SetupRuntime.java]: As far as I understand, initialize() method is used just to initialize some properties and data in the runtime. setup() method is designed to do some work using some objects, which are available after validate() method call. For example a database connection. It is established in validate() method, and then can be used in runtime for other pusrposes. Such purposes could be:
-creating a table in the database;
-cleaning data in database;
-creating some special records in the table;
-downloading/uploading some files to remote servers.
It shouldn't be done before we validate the connection. And also shouldn't be done during validation. it is something that should be processed separately after we initialize() and validate() the runtime.

File: components-api/src/main/java/org/talend/components/api/component/SetupComponent.java (hunk 3 above)
Comment (chmyga) [inline components-api/src/main/java/org/talend/components/api/component/SetupComponent.java]: During the discussion with Sebastien, he suggested we could have different method names, so that the Component could return different runtimes at the same time. Also, as I understand, it should be possible to get and run the setup runtime before all the other component tasks are performed. It could be used for preprocessing some data. For example, creating some table, or deleting some data from it. It should be run before the "main" runtime is processed.

File: components-api/src/main/java/org/talend/components/api/component/runtime/Runtime.java (hunk 1 above)
Comment (sgandon) [inline components-api/src/main/java/org/talend/components/api/component/runtime/Runtime.java]: fine by me.

File: components-api/src/main/java/org/talend/components/api/component/runtime/SetupRuntime.java (hunk 2 above)
Comment (sgandon) [inline components-api/src/main/java/org/talend/components/api/component/runtime/SetupRuntime.java]: I agree that this method has a different semantic thatn initialise that is supposed to store the Properties for the runtime and the validate is supposed to validate the component before the process and may even stop the run of the job whereas the *setup* mean it the job is already running and this method is called after the start of the job and everything has been validated.
So we should keep this.

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/294/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/294/)

File: components-api/src/main/java/org/talend/components/api/component/SetupComponent.java (hunk 3 above)
Comment (sgandon) [inline components-api/src/main/java/org/talend/components/api/component/SetupComponent.java]: As mentioned in the [Jira](https://jira.talendforge.org/browse/TCOMP-234?focusedCommentId=433550&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-433550) SourceOrSink provide methods for getting the schema and schema names so it can not be a generalised *Runtime* interface.
@francisu  I also talked with you of the limitation of the same method name *getRuntime()* returning different object type that was a problem for implementing multiple kind of components into one implem. And you are right we could have a *getRuntime()* in ComponentDefinition, but this brings another question that we need to discuss. if you have a single method that may return different runtime (one source or one sink for instance) how and who will make the choice, how to know the required runtime is for the input comp or the output component.
I would suggest that we keep Dmytro suggestion just for this PR and talk about this on skype to resolve this issue which is a separate matter.

File: components-api/src/main/java/org/talend/components/api/component/runtime/SetupRuntime.java (hunk 2 above)
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/runtime/SetupRuntime.java]: I'm still not in agreement here. The distinction between the 3 methods is still muddy to me. The distinction between initialize() and validate() is clear (as originally defined in GDF), initialize is something entirely local, and validate() will have the component engage with the thing that it's working with (and make sure everything is OK). This is called after the job is started. I don't see a clear distinction between what would go in validate() as opposed to setup(). All of the examples that Dmytro mentioned above would nicely belong in validate() in my estimation; and if any of those things failed, it can report a clear error that the component failed to initialize because of some environmental problem. The connection could be validated as the first thing here. 

@sgandon if you look at this from a DI perspective, the validate() and setup() methods would be called at the same time (in the DI begin method). There is nothing that would happen between the two. 

A legitimate reason for a 3rd method (setup) is that something must happen between validate() and setup(), but that's not the case. I think the GDF people thought though this well, including the names of the methods, and we should just go with their model. Adding this additional method is unnecessary and confusing.

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/319/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/319/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/322/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/322/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/325/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/325/)

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/326/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/326/)

**Comment by RyanSkraba** [comment]
Hello (not the reviewer, just interested)!  This looks like it accurately reflects our long discussion -- in any case, it looks good to me.

A minor objection: I would have preferred to see a `IllegalSpecificationException` or something other than `ValidationResult`.  Exceptions are made for error handling... and "validate" is ruined for us as a verb because of the weird Beam semantics on its validate() method!

But in the end, it doesn't matter _that_ much and I'm fine either way.

**Comment by gonchar-ivan** [comment]
It looks good to me.

File: components-api/src/main/java/org/talend/components/api/component/runtime/ComponentRuntime.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ * {@link PreProcessRuntime}.
+ */
+public interface ComponentRuntime {
+
+    /**
+     * Gets the ComponentProperties associated with this runtime class in order to configure it. This class also checks
+     * that all required properties are properly set and coherent. No integrity check must be done like checking a
+     * remote connection or checking a file exists here. This method shall be called before the job or flow is executed
+     * and shall prevent execution if the ValidationResult has an ERROR status.
+     *
```
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/runtime/ComponentRuntime.java]: You should also make it clear here that this method is expected to be called in a distributed environment before the runtime is serialized and sent to other nodes for execution.

File: components-api/src/main/java/org/talend/components/api/component/runtime/SourceOrSink.java [hunk 5]
```diff
@@ -15,42 +15,30 @@
...
- * This allows the {@code SourceOrSink} instance created in this "main program" to be sent (in serialized form) to remote worker
- * machines and reconstituted for each batch of elements being processed. A {@code SourceOrSink} can have instance variable state,
- * and non-transient instance variable state will be serialized in the main program and then deserialized on remote worker
- * machines.
+ * A {@code SourceOrSink} may be passed between partititions for distributed operation and therefore must be
+ * {@code Serializable}. This allows the {@code SourceOrSink} instance created in this "main program" to be sent (in
+ * serialized form) to remote worker machines and reconstituted for each batch of elements being processed. A
+ * {@code SourceOrSink} can have instance variable state, and non-transient instance variable state will be serialized
+ * in the main program and then deserialized on remote worker machines.
  *
```
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/runtime/SourceOrSink.java]: Actually these comments should be moved to ComponentRuntime instead of here. This clearly states what's going on and now that's at the level of ComponentRuntime.

File: components-api/src/main/java/org/talend/components/api/component/runtime/SourceOrSink.java [hunk 6]
```diff
@@ -61,4 +49,9 @@
      */
     Schema getEndpointSchema(RuntimeContainer container, String schemaName) throws IOException;
 
+    /**
+     * Checks that this source or sink is valid, before it can be used. This will typically make a connection and return
+     * the results of the connection establishment.
+     */
+    ValidationResult validate(RuntimeContainer container);
```
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/runtime/SourceOrSink.java]: This will be called in the same process where the runtime will actually be executed.

File: components-api/src/main/java/org/talend/components/api/component/runtime/PreProcessRuntime.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ * components which need some pre-processing before performing the actual job.
+ */
+public interface PreProcessRuntime extends ComponentRuntime {
+
+    /**
+     * execute some code before the job or flow is started
+     * 
+     * @throw an RuntimeException if anything fails during the run of this method.
+     */
+    void runPreProcess();
```
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/runtime/PreProcessRuntime.java]: The semantics of this is not clear, especially in a distributed environment. From the previous conversations it seems that this is guaranteed (how?) to be called only once for each entire system to do one-time actions. But is it to be called in the process where this runtime instance will eventually execute? Is this intended for components where no further runtime activity is expected? This all needs to be made clear in the javadoc for the method.

**Comment by francisu** [comment]
@RyanSkraba actually on reflection, I think that ValidationResult is an unfortunate name for that class, as it is really only used to contain error information (beyond just an exception). I don't have a better name though. Maybe something like StatusInformation. Changing the name of that class will hit a lot of things (the Studio code depends on it), so I'm not sure it's worth it at this point, and it certainly should not be considered as part of this PR.

**Comment by francisu** [comment]
@RyanSkraba actually on reflection, I think that ValidationResult is an unfortunate name for that class, as it is really only used to contain error information (beyond just an exception). I don't have a better name though. Maybe something like StatusInformation. Changing the name of that class will hit a lot of things (the Studio code depends on it), so I'm not sure it's worth it at this point, and it certainly should not be considered as part of this PR.

**Comment by francisu** [comment]
@chmyga I'm sorry this is taking so long, I know it must be frustrating. But it really must be right as this is something we are going to live with for a long time. I'm going to be on European time for the next couple of weeks, so we should be able to finish this in the next couple of days, as we can talk real time if necessary, and I will try and be responsive.

File: components-api/src/main/java/org/talend/components/api/component/runtime/SourceOrSink.java (hunk 5 above)
Comment (chmyga) [inline components-api/src/main/java/org/talend/components/api/component/runtime/SourceOrSink.java]: I'm not sure we should move them, as ComponentRuntime is not Serializable and it might not be transferred between nodes in distributed environment(PreProcessRuntime is an example of driver-node only runtime).

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/363/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/363/)

File: components-api/src/main/java/org/talend/components/api/component/runtime/SourceOrSink.java (hunk 5 above)
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/runtime/SourceOrSink.java]: OK, so PreProcessRuntime solves the problem of TCOMP-235, having code that's executed once before the execution of the component starts on the driver node. I think the "PreProcessRuntime" name could be changed to describe that better. And from what I understand, this could be used for any distributed component which would also have a regular Source/Sink. I would prefer the interface be called something like ComponentDriverInitialization, and the method be called initializeAtDriver() to make it very clear what's going on. The flow of execution and the reason this is not serializable should be clearly described in the Javadoc.

File: components-api/src/main/java/org/talend/components/api/component/runtime/ComponentRuntime.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+     * This method is expected to be called in a distributed environment on driver node before the runtime is serialized
+     * and sent to worker nodes for execution.
+     * </p>
+     * 
+     * @param container container instance to get access or share data with the running container.
+     * @param properties used to configure this runtime class.
+     * @return ValidationResult status of the validation, if the status is ERROR then the flow or job shall be stopped.
+     */
+    ValidationResult initialize(RuntimeContainer container, ComponentProperties properties);
+
```
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/runtime/ComponentRuntime.java]: This should be serializable to satisfy the requirements of TCOMP-234, which allows a distributed component without connectors. It has to be serializable because the properties are to be moved to the cluster nodes. This should not be a conflict with PreProcessRuntime, which itself does not need to be serializable.

Thanks for the improvement of the javadoc, it's very clear now!

File: components-api/src/main/java/org/talend/components/api/component/runtime/PreProcessRuntime.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+     * runtime process. In case of components with connectors, it will be called before serializing and sending runtimes
+     * to worker nodes.
+     * <p>
+     * In case of components without connectors, no other work should be performed for such runtime, only this method
+     * will be called after runtime initialization.
+     * </p>
+     * 
+     * @throw an RuntimeException if anything fails during the run of this method.
+     */
+    void runPreProcess();
```
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/runtime/PreProcessRuntime.java]: Thanks for the javadoc changes, I'm happy with them. As stated below, I would like the interface and method names to change.

File: components-api/src/main/java/org/talend/components/api/component/runtime/SourceOrSink.java (hunk 5 above)
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/runtime/SourceOrSink.java]: This should be moved to ComponentRuntime, as that must be serializable (see previous comment).

Comment (chmyga) [inline components-api/src/main/java/org/talend/components/api/component/runtime/SourceOrSink.java]: Maybe we could name the method in some other way, like "runAtDriver()"? In my opinion, developers would think that only some initialization process should be done in method, which includes word"initialize" in its name. But this method would include any different process for components without connectors, for example. So I think the name should be more common than "initialize...". But that is only my opinion. It can always be discussed.

File: components-api/src/main/java/org/talend/components/api/component/runtime/ComponentRuntime.java (hunk 8 above)
Comment (chmyga) [inline components-api/src/main/java/org/talend/components/api/component/runtime/ComponentRuntime.java]: PreProcessRuntime should be used for requirements described in TCOMP-234. And I just don't know how will the components without connectors be run on cluster nodes. They aren't Input or Output components, so there's no data passed through them. Maybe I am missing something or we missed something during our discussions?

File: components-api/src/main/java/org/talend/components/api/component/runtime/SourceOrSink.java (hunk 5 above)
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/runtime/SourceOrSink.java]: +1 for runAtDriver()

File: components-api/src/main/java/org/talend/components/api/component/runtime/ComponentRuntime.java (hunk 8 above)
Comment (francisu) [inline components-api/src/main/java/org/talend/components/api/component/runtime/ComponentRuntime.java]: A configuration component for example is distributed but not a connector. It would not use PreProcessRuntime, but also have no connections. It would communicate with other components in the worker node using the methods in RuntimeContainer.

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/379/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/379/)

File: components-api/src/main/java/org/talend/components/api/component/runtime/SourceOrSink.java [hunk 10]
```diff
@@ -15,42 +14,23 @@
...
-     * Initialize based on the specified properties. This will typically store the {@link ComponentProperties} in the
-     * object.
-     */
-    void initialize(RuntimeContainer container, ComponentProperties properties);
-
-    /**
-     * Checks that this source or sink is valid, before it can be used. This will typically make a connection and return
-     * the results of the connection establishment.
-     */
-    ValidationResult validate(RuntimeContainer container);
```
Comment (sgandon) [inline components-api/src/main/java/org/talend/components/api/component/runtime/SourceOrSink.java]: Why did you remove the validate method here, as we need it for beam and use it today to do the integrity checks ?

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/395/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/395/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/396/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/396/)

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/397/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/397/)

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/403/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/403/)

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/404/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/404/)

**Comment by build-talend** [comment]
Pull request test result: [failure](https://ci-common.datapwn.com/job/components_pull-request_master/405/) [:x:](https://ci-common.datapwn.com/job/components_pull-request_master/405/)

**Comment by build-talend** [comment]
Pull request test result: [success](https://ci-common.datapwn.com/job/components_pull-request_master/408/) [:star:](https://ci-common.datapwn.com/job/components_pull-request_master/408/)
